Unify element activation, element death and moving interface

hugary1995 opened the pull request:
Generalize the existing ActivateElementUserObjects, add ElementSubdomainModifiers, and fix some parallel bugs.

close #17100

<!--
INCLUDE THE FOLLOWING IN THE PR DESCRIPTION
- Explain relevant design information for your change.
- Follow the [Coding Standards](http://mooseframework.org/wiki/CodeStandards/).
- Submit or improve [Test Cases](http://mooseframework.org/wiki/MooseTraining/testing/).
- Reference a specific issue, place "refs #<issue>" or "closes #<issue>" (e.g., #closes #1234).
-->

hugary1995 committed:
Unify element activation, element death and moving interface

Generalize the existing ActivateElementUserObjects, add ElementSubdomainModifiers, and fix some parallel bugs.

close #17100

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/689843/) on 3cdb20c wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/17101/site)

This comment will be updated on new commits.

hugary1995 commented:
@fdkong @dschwen Could you please review this since you reviewed the original #16008 PR about element activation?

The new UO changes some of the logic (and names of several parameters, variables, etc) in the original UO so that element activation can be generalized to element death and moving interface.

The original regression tests didn't capture several parallel nodeset/sideset generation issues. I fixed those and added more regression tests.

The [documentation](https://mooseframework.inl.gov/docs/PRs/17101/site/source/userobject/CoupledVarThresholdElementSubdomainModifier.html) should be helpful. It also explicitly mentions what we can do with this UO.

This PR will allow me to couple the element subdomain modifier with XFEM, to get smooth boundary/interface instead of the zig-zag boudaries. @bwspenc the XFEM element subdomain modifier will have to wait for this one to be merged first.

hugary1995 commented on framework/src/userobject/ElementSubdomainModifier.C:
```diff
@@ -0,0 +1,347 @@
+    // Apply initial condition for the newly moved elements and boundary nodes
+    buildMovedElemsRange();
+    buildMovedBndNodesRange();
+    _fe_problem.projectInitialConditionOnCustomRange(movedElemsRange(), movedBndNodesRange());
```
I don't know if I need to also project initial conditions on displaced elements. I am only projecting ICs on undisplaced elements.

hugary1995 replied:
I haven't seen any issues so far without initializing solutions on the displaced mesh. But as you said there maybe corner cases I'm not aware of. I'll just rely on `syncSolutions()` for now.

hugary1995 replied:
Hold on, I actually got segfault if I call `_displaced_problem->syncSolutions()`. Not sure why.

hugary1995 commented on test/tests/userobjects/element_subdomain_modifier/block_restricted.i:
```diff
@@ -0,0 +1,70 @@
+    threshold = 0
+    subdomain_id = 1
+    moving_boundary_name = moving_boundary
+    execute_on = 'INITIAL TIMESTEP_BEGIN'
```
There is still a bug I need to fix: if I don't add INITIAL here, the "moving_boundary" nodeset/sideset won't be created.

hugary1995 replied:
Update: this is due to the current limitation that libmesh doesn't output empty nodeset/sideset. This will be fixed after [https://github.com/libMesh/libmesh/pull/2871](https://github.com/libMesh/libmesh/pull/2871).

loganharbour replied:
Do you plan on waiting for a libmesh update to get this in or do you want to get it in first? If the latter - make an issue and be sure to reference the issue on the test.

hugary1995 replied:
There is no time pressure on this, so I'm fine either way. I prefer to get this in first (and a regression test for the empty nodeset/sideset after a libmesh update). But I'm open to the other option if you think it is better.

hugary1995 replied:
There is also a possibility that a libmesh update comes before this PR is ready to merge.

loganharbour replied:
I have no preference. Just if you choose to get it in before the libmesh update, please create an issue and reference it in the test/test spec

hugary1995 replied:
Okay I created an issue and referenced it.

dewenyushu reviewed (dismissed):
Looks great!

This PR is a better and generalized version of `ActivateElementsUserObjectBase`. Maybe we can later convert the tests from the old element activation code and combine these two.

fdkong commented on framework/include/userobject/ElementSubdomainModifier.h:
```diff
@@ -0,0 +1,95 @@
+
+  ConstElemRange & movedElemsRange() const { return *_moved_elems_range.get(); }
+
+  ConstBndNodeRange & movedBndNodesRange() const { return *_moved_bnd_nodes_range.get(); }
```
I am curious about if we are able to de-reference the object directly from the smart pointer (instead of raw pointer)?

loganharbour replied:
You can remove the `.get()`. The dereference operator for a smart pointer returns a reference to the underlying object.

hugary1995 replied:
You are right, I removed the `.get()`s.

fdkong commented on framework/include/userobject/ElementSubdomainModifier.h:
```diff
@@ -0,0 +1,95 @@
+  DisplacedProblem * _displaced_problem;
+
+private:
+  void setMovingBoundaryName(MooseMesh & mesh);
```
Do you mind adding comments for all of the following functions?

hugary1995 replied:
Added.

fdkong commented on framework/src/userobject/ElementSubdomainModifier.C:
```diff
@@ -0,0 +1,353 @@
+#include "libmesh/dof_map.h"
+
+#include "libmesh/parallel_ghost_sync.h"
+#include "libmesh/mesh_communication.h"
```
We may not need this header if it is already a parallel object

hugary1995 replied:
Removed.

fdkong commented on framework/src/userobject/ElementSubdomainModifier.C:
```diff
@@ -0,0 +1,353 @@
+ElementSubdomainModifier::setMovingBoundaryName(MooseMesh & mesh)
+{
+  // We only need one boundary to modify. Create a dummy vector just to use the API.
+  const std::vector<BoundaryID> boundary_ids = mesh.getBoundaryIDs({{_moving_boundary_name}}, true);
```
double `{` ?

hugary1995 replied:
I think I got min gcc compiler error if I don't use double `{`.

fdkong commented on framework/src/userobject/ElementSubdomainModifier.C:
```diff
@@ -0,0 +1,353 @@
+{
+  // We only need one boundary to modify. Create a dummy vector just to use the API.
+  const std::vector<BoundaryID> boundary_ids = mesh.getBoundaryIDs({{_moving_boundary_name}}, true);
+  _moving_boundary_id = boundary_ids[0];
```
`mooseAssert(boundary_ids.size==1, ".... ... ")`

hugary1995 replied:
Added.

fdkong commented on framework/src/userobject/ElementSubdomainModifier.C:
```diff
@@ -0,0 +1,353 @@
+    // Current element ID, used to index both the element on the displaced and undisplaced meshes.
+    dof_id_type elem_id = _current_elem->id();
+    // Change the element's subdomain
+    Elem * elem = _mesh.elemPtr(elem_id);
```
`elem` will be different from `_current_elem`?  That being said, can we directly use `_current_elem` instead of `elem` in the following several places?

hugary1995 replied:
But `_current_elem` is `const`, so I can't use it to directly change the subdomain ID.

fdkong commented on framework/src/userobject/ElementSubdomainModifier.C:
```diff
@@ -0,0 +1,353 @@
+  }
+
+  // Initialize stateful material properties for the newly activated elements
+  _fe_problem.initElementStatefulProps(movedElemsRange());
```
I feel like it is better to  change "`moved`" to `activated`.  `activated` makes more sense for in this context

hugary1995 replied:
I don't think so. The elements here really just changed subdomain IDs. `activated` would make perfect sense if we use this object for element activation, but would be misleading in the context of element deletion and moving interface.

fdkong commented on framework/src/userobject/ElementSubdomainModifier.C:
```diff
@@ -0,0 +1,353 @@
+
+  NumericVector<Number> & current_solution = *sys.system().current_local_solution;
+  NumericVector<Number> & old_solution = sys.solutionOld();
+  NumericVector<Number> & older_solution = sys.solutionOlder();
```
`ElementSubdomainModifier` has to use with a transient executioner?  If not, we need to check if we need to use old and older solutions

hugary1995 replied:
How do I check that?

fdkong replied:
You may always need a transient? The element activation won't happen for a steady simulation. 

If this assumption is right, we could merge the code as is

hugary1995 replied:
This is supposed to work in steady as well. I fixed it.

fdkong commented on framework/src/userobject/ElementSubdomainModifier.C:
```diff
@@ -0,0 +1,353 @@
+    older_solution.set(dof, current_solution(dof));
+  }
+
+  current_solution.close();
```
We do not need `close` for current_solution since we do not really change values

hugary1995 replied:
Right.

fdkong commented on framework/src/userobject/ElementSubdomainModifier.C:
```diff
@@ -0,0 +1,353 @@
+  for (auto dof : dofs)
+  {
+    old_solution.set(dof, current_solution(dof));
+    older_solution.set(dof, current_solution(dof));
```
Why `older_solution` does not take solution from `old_solution` ?

hugary1995 replied:
So I am only setting ICs when `apply_ic = true`. That is meant for element activation/deletion, which implies that these dofs just moved from an inactive domain to an active domain, or vice versa. Therefore, I am applying the IC as their current/old/older solutions -- mimicking an actual IC.

hugary1995 replied:
Let me know whether you think this makes sense or not.

fdkong replied:
This makes sense if it is applied to only "new" elements

fdkong commented on framework/src/userobject/ThresholdElementSubdomainModifier.C:
```diff
@@ -0,0 +1,65 @@
+      break;
+  }
+
+  return criterion_met ? _subdomain_id : _complement_subdomain_id;
```
`ThresholdElementSubdomainModifier` will handle only two subdomains?

hugary1995 replied:
Right now yes. Actually if the user needs more than one thresholds (more than two subdomains to switch between), they can add multiple `ThresholdElementSubdomainModifer`s in their input file.

fdkong reviewed (changes requested):
Looks very good for me.

dschwen commented:
What's the deal with the empty jpg files added in this PR?

hugary1995 commented:
> What's the deal with the empty jpg files added in this PR?

I don't know. The images are actually not empty if you check them out. I guess it's a github issue?

hugary1995 commented:
@fdkong could you take another look at it? When I said "there is no time pressure on this" I wasn't expecting this to take two weeks. I have a report due in April. Thanks.

hugary1995 commented:
@fdkong I followed your suggestions and added a regression test for the steady case.

dschwen commented on framework/src/userobject/CoupledVarThresholdElementSubdomainModifier.C:
```diff
@@ -0,0 +1,41 @@
+  Real avg_val = 0;
+
+  for (unsigned int qp = 0; qp < _qrule->n_points(); ++qp)
+    avg_val += _v[qp];
```
should this be multiplied by `_JxW[_qp] * _coord[_qp]`?

hugary1995 replied:
Good catch! I fixed that. @dewenyushu this will probably improve some of your AM results.

fdkong replied:
It is indeed a good catch!!

dewenyushu replied:
Oops I guess so! I had trouble with element activation using coupled variable. Maybe this is one of the reasons! Nice catch!

hugary1995 committed:
use element average

fdkong commented on framework/src/userobject/ElementSubdomainModifier.C:
```diff
@@ -0,0 +1,360 @@
+{
+  // Don't do anything if this is a steady simulation
+  if (!sys.hasSolutionState(1))
+    return;
```
I like it!

fdkong reviewed (approved):
Looks great

loganharbour commented:
@hugary1995 this is failing rather painfully almost everywhere in parallel: https://civet.inl.gov/event/65489/

Are you able to look into this asap or do we need to?

hugary1995 commented:
Sure! I'll look at it ASAP!

hugary1995 commented:
@loganharbour I can't figure it out any time soon. Can we revert it?
